qualification/report/generate_pdf.py

```diff
@@ -835,13 +835,42 @@ def document_from_json(input_data: str | list) -> Document:
     return doc
 
 
+def test_image_commit(input_data: str | list) -> str:
+    """Extract the git commit ID of the test image from the test results.
+
+    Parameters
+    ----------
+    input_data: str | list
+        Either the list of parsed JSON entries, or a path to the report file
+        generated by :samp:`pytest --report-log={filename}`.
+
+    Returns
+    -------
+    commit_id: str
+        The git commit ID of the test image.
+    """
+    if not isinstance(input_data, list):
+        result_list = []
+        with open(input_data) as fp:
+            for line in fp:
+                result_list.append(json.loads(line))
+    else:
+        result_list = input_data
```

Comment:
This whole piece is copy-pasted. I think rather make a function that opens and parses the file, and then just pass a list to both `test_image_commit` and `document_from_json`.